### PR TITLE
Adds upload option 'dirpage'

### PR DESCRIPTION
**Pull request by hannwong**
Combines options 'dir' and 'page'.

Allows the reuse of filenames for uploaded files. Best of both worlds. Allows the consolidation of uploads in folder uploads, and also creates subfolder structure that mirrors that provided by option page.

### CODE REVIEW

**Comment by bartkamphorst** [comment]
All checks now pass on Travis. @dometto What are your thoughts on this PR?

**Comment by hannwong** [comment]
@bartkamphorst Hi! Is gollum 5.x now nearing release? :D Yup, I've been using that `dirpage` option with much success. U can merge this PR. Gimme a holler if u see me break anything.

File: lib/gollum/app.rb [hunk 1]
```diff
@@ -583,10 +583,16 @@ def commit_message
     end
 
     def find_upload_dest(path)
-      settings.wiki_options[:allow_uploads] ?
-          (settings.wiki_options[:per_page_uploads] ?
-              "#{path}/#{@name}".sub(/^\/\//, '') : 'uploads'
-          ) : ''
+      if not settings.wiki_options[:allow_uploads]
```
Comment (dometto) [inline lib/gollum/app.rb]: Please use `!` rather than `not` for style consistency

File: lib/gollum/app.rb [hunk 2]
```diff
@@ -583,10 +583,16 @@ def commit_message
...
+        return ''
+      end
+
+      dest = settings.wiki_options[:per_page_uploads] ?
+               "#{path}/#{@name}".sub(/^\/\//, '') : 'uploads'
+      dest = settings.wiki_options[:per_page_uploads_cons] ?
```
Comment (dometto) [inline lib/gollum/app.rb]: Don't think the ternary operator is necessary here. Perhaps just:
`dest =  ::File.join([settings.wiki_options['page_file_dir'], 'uploads/', dest].compact) if settings.wiki_options[:per_page_uploads_cons]`

**Comment by dometto** [comment]
No objections apart from the line comments, nice to have the extra feature. It looks like we don't have tests for file uploads yet. It would be nice to have this, also for the new option, if you feel like adding them, @hannwong!  otherwise, I think what would be nice to have is an more elaborate description of the different upload modes (with examples of the resulting directory structures) to go into the wiki.

**Comment by hannwong** [comment]
> It looks like we don't have tests for file uploads yet

@dometto I'll add those tests. I should.

File: lib/gollum/app.rb (hunk 2 above)
Comment (dometto) [reply lib/gollum/app.rb]: Or keep the ternary operator, but ditch the assignment and return statement.
`settings.wiki_options[:per_page_uploads_cons] ? ::File.join([settings.wiki_options['page_file_dir'], 'uploads/', dest].compact) : dest` (and remove the `return` on 595)

**Comment by bartkamphorst** [comment]
@hannwong Any update with regard to adding tests and fixing the code suggestions from @dometto ? I'd love to incorporate this feature into `5.x`.

**Comment by bartkamphorst** [comment]
ping @hannwong .

**Comment by bartkamphorst** [comment]
This PR is now superseded by https://github.com/gollum/gollum/pull/1344.

**Pull request closed without merging**